core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,203 @@
+   */
+  def shutdown(): Unit
+
+  /**
+   * Views this `Executor` as a Scala `ExecutionContext`.
+   */
+  lazy val asExecutionContext: ExecutionContext =
+    new ExecutionContext {
+      override def execute(r: Runnable): Unit =
+        if (!submit(r)) throw new RejectedExecutionException("Rejected: " + r.toString)
+
+      override def reportFailure(cause: Throwable): Unit =
+        cause.printStackTrace
+    }
+}
+
+object Executor extends Serializable {
+  sealed abstract class Type extends Product with Serializable
+  object Type extends Serializable {
+
+    /**
+     * An executor optimized for synchronous tasks, which yield
+     * to the runtime infrequently or never.
+     */
+    final case object Synchronous extends Type
+
+    /**
+     * An executor optimized for asynchronous tasks, which yield
+     * frequently to the runtime.
+     */
+    final case object Asynchronous extends Type
+  }
+
+  /**
+   * Creates a new default executor of the specified type.
+   */
+  final def newDefaultExecutor(tpe: Type): Executor = tpe match {
+    case Type.Synchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Int.MaxValue
+        val maxPoolSize   = Int.MaxValue
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio-blocking", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+
+    case Type.Asynchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Runtime.getRuntime.availableProcessors() * 2
+        val maxPoolSize   = corePoolSize
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+  }
+
+  /**
+   * Creates an `Executor` from a Scala `ExecutionContext`. You must manually
+   * supply the concurrency level, since `ExecutionContext` cannot
+   * provide this detail.
+   */
+  final def fromExecutionContext(n: Int, ec: ExecutionContext): Executor =
+    new Executor {
+      import java.util.concurrent.atomic.AtomicLong
+
+      val _enqueuedCount = new AtomicLong()
+      val _dequeuedCount = new AtomicLong()
+
+      def concurrency: Int = n
+
+      def capacity: Int = Int.MaxValue
+
+      def size: Int = (enqueuedCount - dequeuedCount).toInt
```

Comment:
It looks like `size` would then return the number of tasks that have not yet been started or have not yet completed. Do we also want the number of tasks that are just sitting in the queue and have not been picked up by a thread yet?

Reply:
/cc @artempyanykh 

I was trying to make this interface look like Artem's queue. It seems this is a bit ambiguous. Having both pieces of information could be useful.

Reply:
I'm thinking whether we actually need both pieces of information? I don't have full clarity on scheduling algo v2, but assume that scheduling decisions should be based on avg. rates rather than point in time measurements.

If we increment `_dequeuedCount` before `runnable.run()` then `size` will give us the number of tasks *waiting in the queue*, which is the metric that I'd be more interested in. 

The avg `size` can grow only when all threads are saturated over a period of time (i.e. `# currently running tasks == concurrency`) and that's when we have to apply scheduling magic.

Reply:
For the case when the concurrency is unbounded (synchronized queue pools for blocking ops) it seems enough to monitor `avg. concurrency` as nothing will be queued. So, looks like the pair `(avg. size, avg. concurrency)` covers what we need.

Reply:
I'm sorry for being verbose, but I can't stop arguing with myself 😄 @jdegoes do you want a single interface for both types of executors? It just seems that
1. for `Type.Asynchronous` the # of items waiting in the queue is the most interesting metric,
2. but for `Type.Synchronous` we need to know the # of currently running tasks as there's no queueing actually going on.

Reply:
For synchronous, do we need to know anything? If we yield, it just goes onto a new thread, which is inefficient. Basically, synchronous should never yield.

Reply:
Btw I am thinking to call them `Unyielding` and `Yielding` instead of `Synchronous` and `Asynchronous`.

Reply:
`Relentless` / `Submissive` 😆

Reply:
Dom / Sub. If we’re going there. 😆

Reply:
Approved 💯

Reply:
What about the Drupal kink police?

OH RIGHT. This is Scalaz. 😆

Reply:
I’m probably blissfully ignorant here ;-)

Reply:
For scheduling purposes, the only thing that matters is how frequently the program yields to the runtime system. So a CPU-bound application that does all its work inside `sync` needs to go on the `Unyielding` pool, otherwise it will clog the async tasks.

Reply:
I like `Yielding`/`Unyielding` :+1: Although, I'm not 100% convinced that tasks submitted to `Unyielding` should jump onto a thread right away. 

Some queueing can be beneficial to balance the # of threads in the `Unyielding` executor. Maybe a `ForkJoinPool` would be a better default?

Reply:
I'm open to revise the defaults for `Unyielding`, but let's do that in a followup PR (I have a feeling lots of people will have input on that, as for the defaults for the async pool).